### PR TITLE
Lint for sorted regions in Rust code

### PR DESCRIPTION
**Pull request by antiguru**
This adds a lint for sorted regions in Rust files. Regions are started with

    // keep-sorted-start

and end with

    // keep-sorted-end

The contents between these markers are checked to be sorted. Note that this lint is not able to understand Rust code; it only checks the contents line-by-line.

Signed-off-by: Moritz Hoffmann <mh@materialize.com>

### CODE REVIEW

File: misc/lint/keep-sorted.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+if [[ $# -lt 1 ]]; then
+    echo "usage: $0 <file>" >&2
+    exit 1
+fi
+errors=0
+
+for file in "$@"; do
+    if [[ ! -f "$file" ]]; then
+        echo "lint: $(red error:) trailing-newline: internal error: $file is not a file" >&2
```
Comment (benesch) [inline misc/lint/keep-sorted.sh:27]:
```suggestion
        echo "lint: $(red error:) keep-sorted: internal error: $file is not a file" >&2
```

File: misc/lint/keep-sorted.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    diff "$file" <(gawk '/\/\/ keep-sorted-start/ {f=1; delete a}
+      /\/\/ keep-sorted-end/ {f=0; n=asort(a); for (i=1;i<=n;i++) print a[i]}
+      !f
+      f{a[$0]=$0}' "$file") && error=$? || error=$?
+    if [ "$error" -gt 1 ]; then
+        echo "lint: $(red error:) diff internal error $error"
+        exit "$error"
+    elif [ "$error" -eq 1 ]; then
+        echo "lint: $(red error:) $file not sorted"
```
Comment (benesch) [inline misc/lint/keep-sorted.sh:39]:
```suggestion
        echo "lint: $(red error:) keep-sorted blocks in $file are not correctly sorted"
```

File: misc/lint/keep-sorted.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+for file in "$@"; do
+    if [[ ! -f "$file" ]]; then
+        echo "lint: $(red error:) trailing-newline: internal error: $file is not a file" >&2
+        exit 1
+    fi
+
+    diff "$file" <(gawk '/\/\/ keep-sorted-start/ {f=1; delete a}
+      /\/\/ keep-sorted-end/ {f=0; n=asort(a); for (i=1;i<=n;i++) print a[i]}
+      !f
+      f{a[$0]=$0}' "$file") && error=$? || error=$?
```
Comment (benesch) [inline misc/lint/keep-sorted.sh:34]: We can't count on `gawk` being installed, I don't think. It's not installed by default on macOS. I also think GNU awk is usually just called `awk` on Linux right?

Does this not work with BSD awk?

Comment (antiguru) [reply misc/lint/keep-sorted.sh:34]: Indeed, this used GNU awk extensions. Changing the PR to not use any extensions. Thanks for pointing this out!

File: misc/lint/keep-sorted.sh (hunk 2 above)
Comment (benesch) [reply misc/lint/keep-sorted.sh:39]: It still uses `gawk` though! Can we change to `awk` now?

**Review by benesch — APPROVED** [review]

File: misc/lint/keep-sorted.sh (hunk 2 above)
Comment (antiguru) [reply misc/lint/keep-sorted.sh]: Now it's not using gnu extensions anymore :) Took me a but to add a `--rewrite` option.

**Comment by antiguru** [comment]
Closing; if we want this feature, it'd need to be more aware of Rust's syntax. This one just operates on lines and breaks as soon as an expression is wrapped.

**Pull request closed without merging**